### PR TITLE
Use descriptor_pool to fix several bugs

### LINKED ISSUE #7: Allow package name and enum types to be used together in proto file

### LINKED ISSUE #11: messages using top-level defined enums pick the wrong library name
Here it's a protobuf test
```proto
syntax = "proto3";

package test;

enum ThisIsAnEnum {
    Nothing = 0;
    Something = 1;
    Anything = 2;
}

message ThisIsAMessage {
    string text = 1;
    ThisIsAnEnum content = 2;
}
```
running
```bash
$ ./run.sh --input test.proto --output .
```
the global-level enum is include in
```solidity
library TEST_PROTO_GLOBAL_ENUMS {

  //enum definition
  // Solidity enum definitions
  enum ThisIsAnEnum {
    Nothing,
    Something,
    Anything
  }
...
```
However the generated code for the message is
```solidity
library TestThisIsAMessage {


  //struct definition
  struct Data {
    string text;
    test.ThisIsAnEnum content;
  }
...
```
You can see that ``Data`` is looking for the enum in the wrong library. ``test`` library does not exist

### LINKED ISSUE #19: Support repeated enum
If you defines a message that has repeated enum fields, compiling the proto file succeeds, but compiling the generated sol file fails.
This happens for both global enums and nested enums.

### PR DESCRIPTION
**Pull request by siburu**
In this PR, I have applied `google.protobuf.descriptor_pool` for code cleanup.
In addition to this, in order to make the code easier to maintain, I added type hints to almost all functions of the code generator.

This PR
- closes #7 
- closes #11 
- closes #19

Signed-off-by: Masanori Yoshida <masanori.yoshida@datachain.jp>

### CODE REVIEW

**Comment by siburu** [comment]
### Test case
```proto
syntax = "proto3";

package pac.man;

enum GlobalEnum {
  GE0 = 0;
  GE1 = 10;
  GE2 = 20;
  GE3 = 30;
}

message UsedStruct {
  GlobalEnum ge = 1;
  string str = 2;
  int32 i32 = 3;
}

message GlobalStruct {
  enum NestedEnum {
    NE0 = 0;
    NE1 = 10;
    NE2 = 20;
    NE3 = 30;
  }
  message NestedStruct {
    GlobalEnum ge = 1;
    NestedEnum ne = 2;
    string str = 3;
    repeated GlobalEnum ge_list = 4;
    repeated NestedEnum ne_list = 5;
    repeated string str_list = 6;
    map<string, GlobalEnum> ge_dict = 8;
    map<string, NestedEnum> ne_dict = 9;
    UsedStruct us = 10;
    repeated UsedStruct us_list = 11;
    map<string, UsedStruct> us_dict = 12;
  }
  GlobalEnum ge = 1;
  NestedEnum ne = 2;
  NestedStruct ns = 3;
  string str = 4;
  repeated GlobalEnum ge_list = 5;
  repeated NestedEnum ne_list  = 6;
  repeated NestedStruct ns_list  = 7;
  repeated string str_list  = 8;
  map<string, GlobalEnum> ge_dict = 10;
  map<string, NestedEnum> ne_dict = 11;
  map<string, NestedStruct> ns_dict = 12;
  UsedStruct us = 13;
  repeated UsedStruct us_list = 14;
  map<string, UsedStruct> us_dict = 15;
}
```

The above proto is successfully compiled to a sol file that is successfully compiled by solc.

File: protobuf-solidity/src/protoc/plugin/gen_util.py [hunk 1]
```diff
@@ -300,14 +291,39 @@ def str_contains(s, token):
   except Exception as e:
     return False
 
-def gen_struct_name_from_field(f):
-  return "".join(map(lambda t: t[:1].upper() + t[1:], f.type_name[1:].split(".")))
+def gen_struct_name(msg: Descriptor) -> str:
+  """Generates PackageNameContainingTypeNameSelfTypeName"""
+  return "".join(map(lambda word: word[:1].upper() + word[1:], msg.full_name.split('.')))
+  #hierarchy = [msg.name]
```
Comment (bluele) [inline protobuf-solidity/src/protoc/plugin/gen_util.py:297]: the comments needed?

File: protobuf-solidity/src/protoc/plugin/gen_util.py [hunk 2]
```diff
@@ -180,28 +177,24 @@
...
 SOLIDITY_PRAGMAS = []
 
 # utils
-def is_map_type(f, nested_type):
-  if f.type_name.endswith("MapFieldEntry") and field_is_repeated(f):
-    return len(list(filter(lambda t: t.name == "MapFieldEntry" and t.options and t.options.map_entry ,nested_type))) > 0
-  return False
+def is_map_type(f: FieldDescriptor) -> bool:
+  return f.message_type and f.message_type.GetOptions().map_entry
+  #if f.message_type and f.message_type.name.endswith("MapFieldEntry") and field_is_repeated(f):
```
Comment (bluele) [inline protobuf-solidity/src/protoc/plugin/gen_util.py:189]: the comments needed?

File: protobuf-solidity/src/protoc/plugin/gen_encoder_constants.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 
 INNER_FIELD_ENCODER_REPEATED_ENUM = """
     {block_begin}
-    int64 _enum_{field};
+    int32 _enum_{field};
```
Comment (bluele) [inline protobuf-solidity/src/protoc/plugin/gen_encoder_constants.py:31]: :+1:

File: protobuf-solidity/src/protoc/plugin/gen_sol.py [hunk 4]
```diff
@@ -1,59 +1,52 @@
...
   if map_fields.strip():
     map_fields = "\n    //non serialized fields" + map_fields
   else:
     map_fields = ""
-  fields = gen_fields(msg, file)
+  fields = gen_fields(msg)
   if (fields or map_fields):
     return (sol_constants.STRUCT_DEFINITION).format(
-      fields = gen_fields(msg, file),
+      fields = gen_fields(msg),
```
Comment (bluele) [inline protobuf-solidity/src/protoc/plugin/gen_sol.py:49]: Can use the return value at L46 instead?

File: protobuf-solidity/src/protoc/plugin/gen_util.py (hunk 1 above)
Comment (siburu) [reply protobuf-solidity/src/protoc/plugin/gen_util.py]: I have removed the comment.

File: protobuf-solidity/src/protoc/plugin/gen_util.py (hunk 2 above)
Comment (siburu) [reply protobuf-solidity/src/protoc/plugin/gen_util.py]: I have removed the comment.

File: protobuf-solidity/src/protoc/plugin/gen_sol.py [hunk 5]
```diff
@@ -76,23 +81,22 @@ def gen_enum_definition(msg, parent_struct_name):
...
-    if nt.options and nt.options.map_entry:
-      pb_nested_struct_name = gen_nested_struct_name(nt, parent_msg, parent_struct_name)
-      if util.gen_struct_name_from_field(f) == pb_nested_struct_name:
+def gen_map_insert_on_store(f: FieldDescriptor, parent_msg: Descriptor) -> str:
+  for nt in parent_msg.nested_types:
+    if nt.GetOptions().map_entry:
+      if f.message_type and f.message_type is nt:
         return ('output._size_{name} = input._size_{name};\n').format(
           name = f.name,
           i = f.number
```
[the remark this answers is not in the document]
Comment (siburu) [reply protobuf-solidity/src/protoc/plugin/gen_sol.py]: Thank you. I've fixed it.

**Review by bluele — APPROVED** [review]
LGTM :+1:

**Pull request merged**